File: packages/contracts-ts/src/scoring.ts
/** Phase 29 scoring and after-action contracts. */

import { z } from 'zod';

import {
  assetIdSchema,
  eventIdSchema,
  evidenceIdSchema,
  hypothesisIdSchema,
  incidentIdSchema,
  proposalIdSchema,
  runIdSchema,
  sequenceSchema,
  utcTimestampSchema,
} from './primitives';
import {
  AFTER_ACTION_VIEW_MODEL_SCHEMA_VERSION,
  DECISION_REVIEW_SCHEMA_VERSION,
  MISSED_EVIDENCE_ITEM_SCHEMA_VERSION,
  RUN_COMPARISON_SCORE_SCHEMA_VERSION,
  RUN_SCORE_SCHEMA_VERSION,
  SCORE_COMPONENT_SCHEMA_VERSION,
  SCORE_EXPLANATION_SCHEMA_VERSION,
  SCORE_EXPORT_ARTIFACT_SCHEMA_VERSION,
  SCORE_PROVENANCE_SCHEMA_VERSION,
  SCORE_RUBRIC_SCHEMA_VERSION,
  VALID_ALTERNATIVE_SCHEMA_VERSION,
} from './versioning';

export const GRADING_ENGINE_VERSION = '1.0.0-phase29' as const;

const schemaVersionCheck = (expected: number) =>
  z
    .number()
    .int()
    .min(1)
    .refine((v) => v === expected);

export const scoreErrorCodeSchema = z.enum([
  'SCORE_INCOMPLETE_RUN',
  'SCORE_RUBRIC_MISSING',
  'SCORE_RUBRIC_INCOMPATIBLE',
  'SCORE_INPUT_TAMPERED',
  'SCORE_CHECKSUM_MISMATCH',
  'SCORE_NOT_FOUND',
  'SCORE_COMPARISON_INVALID',
  'SCORE_EXPORT_FAILED',
  'SCORE_VALIDATION_FAILED',
]);

export const scoreGradeBandSchema = z.enum(['A', 'B', 'C', 'D', 'F']);

export const decisionReviewOutcomeSchema = z.enum([
  'credited',
  'penalized',
  'neutral',
  'restraint_credited',
]);

export const validAlternativeKindSchema = z.enum([
  'counterfactual',
  'valid_response_branch',
  'alternative_decision',
]);

export const scoreExportFormatSchema = z.enum(['json', 'markdown']);

export const scoreExplanationSchema = z
  .object({
    schemaVersion: schemaVersionCheck(SCORE_EXPLANATION_SCHEMA_VERSION),
    ruleId: z.string().min(1).max(128),
    reason: z.string().min(1).max(2048),
    eventIds: z.array(eventIdSchema).default([]),
    evidenceIds: z.array(evidenceIdSchema).default([]),
    decisionIds: z.array(z.string().min(1).max(128)).default([]),
    hypothesisIds: z.array(hypothesisIdSchema).default([]),
    proposalIds: z.array(proposalIdSchema).default([]),
    sequence: sequenceSchema.nullable().optional(),
  })
  .strict();

export const scoreRubricCriterionSchema = z
  .object({
    id: z.string().min(1).max(128),
    label: z.string().min(1).max(256),
    weight: z.number().min(0).max(1),
    description: z.string().max(1024).default(''),
  })
  .strict();

export const scoreRubricSchema = z
  .object({
    schemaVersion: schemaVersionCheck(SCORE_RUBRIC_SCHEMA_VERSION),
    scenarioId: z.string().min(1).max(128),
    scenarioVersion: z.string().min(1).max(64),
    rubricVersion: z.string().min(1).max(64),
    gradingEngineVersion: z.string().min(1).max(64),
    maxScore: z.number().positive(),
    criteria: z.array(scoreRubricCriterionSchema).min(1),
    rubricText: z.string().max(8192).default(''),
    rubricTextHash: z.string().min(1).max(128),
  })
  .strict();

export const scoreComponentSchema = z
  .object({
    schemaVersion: schemaVersionCheck(SCORE_COMPONENT_SCHEMA_VERSION),
    criterionId: z.string().min(1).max(128),
    label: z.string().min(1).max(256),
    weight: z.number().min(0).max(1),
    rawScore: z.number().min(0).max(1),
    weightedContribution: z.number().min(0),
    ruleIds: z.array(z.string().min(1).max(128)).default([]),
    explanations: z.array(scoreExplanationSchema).default([]),
  })
  .strict();

export const scoreProvenanceSchema = z
  .object({
    schemaVersion: schemaVersionCheck(SCORE_PROVENANCE_SCHEMA_VERSION),
    runId: runIdSchema,
    scenarioId: z.string().min(1).max(128),
    scenarioVersion: z.string().min(1).max(64),
    rubricVersion: z.string().min(1).max(64),
    gradingEngineVersion: z.string().min(1).max(64),
    inputEventSequenceFrom: sequenceSchema,
    inputEventSequenceTo: sequenceSchema,
    inputChecksum: z.string().min(1).max(128),
    integrityChecksum: z.string().min(1).max(128),
    calculatedAt: utcTimestampSchema,
    fingerprint: z.string().min(1).max(128),
  })
  .strict();

export const decisionReviewSchema = z
  .object({
    schemaVersion: schemaVersionCheck(DECISION_REVIEW_SCHEMA_VERSION),
    decisionId: z.string().min(1).max(128),
    kind: z.enum(['approval', 'rejection', 'modification', 'cancellation', 'agent_recommendation']),
    sequence: sequenceSchema,
    proposalId: proposalIdSchema.nullable().optional(),
    operatorAction: z.string().min(1).max(256),
    agentRecommendation: z.string().max(1024).nullable().optional(),
    availableInfoSummary: z.string().max(2048).default(''),
    availableEventIds: z.array(eventIdSchema).default([]),
    availableEvidenceIds: z.array(evidenceIdSchema).default([]),
    outcome: decisionReviewOutcomeSchema,
    scoreDelta: z.number(),
    ruleIds: z.array(z.string().min(1).max(128)).default([]),
    explanations: z.array(scoreExplanationSchema).default([]),
    usedFutureKnowledge: z.literal(false).default(false),
  })
  .strict();

export const missedEvidenceItemSchema = z
  .object({
    schemaVersion: schemaVersionCheck(MISSED_EVIDENCE_ITEM_SCHEMA_VERSION),
    expectedEvidenceKey: z.string().min(1).max(256),
    eventType: z.string().min(1).max(128),
    assetId: assetIdSchema.nullable().optional(),
    description: z.string().max(1024).default(''),
    discovered: z.boolean(),
    relatedEventIds: z.array(eventIdSchema).default([]),
    relatedEvidenceIds: z.array(evidenceIdSchema).default([]),
    bookmarkSequence: sequenceSchema.nullable().optional(),
  })
  .strict();

export const validAlternativeSchema = z
  .object({
    schemaVersion: schemaVersionCheck(VALID_ALTERNATIVE_SCHEMA_VERSION),
    alternativeId: z.string().min(1).max(128),
    kind: validAlternativeKindSchema,
    authoritative: z.literal(false),
    label: z.string().min(1).max(256),
    description: z.string().max(2048).default(''),
    projectedOverallScore: z.number().min(0),
    scoreDelta: z.number(),
    ruleIds: z.array(z.string().min(1).max(128)).default([]),
    explanations: z.array(scoreExplanationSchema).default([]),
  })
  .strict();

export const runScoreSchema = z
  .object({
    schemaVersion: schemaVersionCheck(RUN_SCORE_SCHEMA_VERSION),
    scoreId: z.string().min(1).max(64),
    runId: runIdSchema,
    incidentId: incidentIdSchema.nullable().optional(),
    overallScore: z.number().min(0),
    maxScore: z.number().positive(),
    grade: scoreGradeBandSchema,
    passed: z.boolean(),
    components: z.array(scoreComponentSchema).min(1),
    provenance: scoreProvenanceSchema,
    decisionReviews: z.array(decisionReviewSchema).default([]),
    missedEvidence: z.array(missedEvidenceItemSchema).default([]),
    validAlternatives: z.array(validAlternativeSchema).default([]),
    coachingText: z.string().max(8192).nullable().optional(),
    coachingAuthoritative: z.literal(false).default(false),
    hiddenCauseId: z.string().max(128).nullable().optional(),
    hiddenCauseLabel: z.string().max(256).nullable().optional(),
    hiddenCauseRevealed: z.boolean().default(false),
  })
  .strict();

export const afterActionBookmarkRefSchema = z
  .object({
    label: z.string().min(1).max(256),
    sequence: sequenceSchema,
    kind: z.enum(['detection', 'evidence', 'decision', 'missed_evidence', 'alternative', 'custom']),
    relatedId: z.string().max(128).nullable().optional(),
  })
  .strict();

export const afterActionViewModelSchema = z
  .object({
    schemaVersion: schemaVersionCheck(AFTER_ACTION_VIEW_MODEL_SCHEMA_VERSION),
    runId: runIdSchema,
    runStatus: z.string().min(1).max(64),
    score: runScoreSchema,
    affectedAssetIds: z.array(assetIdSchema).default([]),
    lessons: z.array(z.string().min(1).max(1024)).default([]),
    scribeReportId: z.string().max(64).nullable().optional(),
    scribeVersionNumber: z.number().int().min(1).nullable().optional(),
    bookmarks: z.array(afterActionBookmarkRefSchema).default([]),
    timelineHighlights: z
      .array(
        z
          .object({
            sequence: sequenceSchema,
            label: z.string().min(1).max(512),
            kind: z.string().min(1).max(64),
            relatedId: z.string().max(128).nullable().optional(),
          })
          .strict(),
      )
      .default([]),
  })
  .strict();

export const runComparisonSchema = z
  .object({
    schemaVersion: schemaVersionCheck(RUN_COMPARISON_SCORE_SCHEMA_VERSION),
    leftRunId: runIdSchema,
    rightRunId: runIdSchema,
    leftScoreId: z.string().min(1).max(64),
    rightScoreId: z.string().min(1).max(64),
    leftOverallScore: z.number().min(0),
    rightOverallScore: z.number().min(0),
    overallDelta: z.number(),
    componentDeltas: z
      .array(
        z
          .object({
            criterionId: z.string().min(1).max(128),
            leftRawScore: z.number().min(0).max(1),
            rightRawScore: z.number().min(0).max(1),
            delta: z.number(),
          })
          .strict(),
      )
      .default([]),
    gradingEngineVersion: z.string().min(1).max(64),
  })
  .strict();

export const scoreExportArtifactSchema = z
  .object({
    schemaVersion: schemaVersionCheck(SCORE_EXPORT_ARTIFACT_SCHEMA_VERSION),
    exportId: z.string().min(1).max(64),
    scoreId: z.string().min(1).max(64),
    runId: runIdSchema,
    format: scoreExportFormatSchema,
    scenarioVersion: z.string().min(1).max(64),
    rubricVersion: z.string().min(1).max(64),
    gradingEngineVersion: z.string().min(1).max(64),
    integrityChecksum: z.string().min(1).max(128),
    contentChecksum: z.string().min(1).max(128),
    createdAt: utcTimestampSchema,
  })
  .strict();

export type ScoreErrorCode = z.infer<typeof scoreErrorCodeSchema>;
export type ScoreGradeBand = z.infer<typeof scoreGradeBandSchema>;
export type ScoreExplanationV1 = z.infer<typeof scoreExplanationSchema>;
export type ScoreRubricV1 = z.infer<typeof scoreRubricSchema>;
export type ScoreComponentV1 = z.infer<typeof scoreComponentSchema>;
export type ScoreProvenanceV1 = z.infer<typeof scoreProvenanceSchema>;
export type DecisionReviewV1 = z.infer<typeof decisionReviewSchema>;
export type MissedEvidenceItemV1 = z.infer<typeof missedEvidenceItemSchema>;
export type ValidAlternativeV1 = z.infer<typeof validAlternativeSchema>;
export type RunScoreV1 = z.infer<typeof runScoreSchema>;
export type AfterActionViewModelV1 = z.infer<typeof afterActionViewModelSchema>;
export type RunComparisonV1 = z.infer<typeof runComparisonSchema>;
export type ScoreExportArtifactV1 = z.infer<typeof scoreExportArtifactSchema>;

File: packages/contracts-python/src/aegis_contracts/scoring.py
"""Phase 29 scoring and after-action contracts."""

from __future__ import annotations

from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from aegis_contracts.errors import ContractErrorCode, ContractValidationError
from aegis_contracts.primitives import (
    AssetId,
    EventId,
    EvidenceId,
    HypothesisId,
    IncidentId,
    ProposalId,
    RunId,
    Sequence,
    UtcTimestamp,
)
from aegis_contracts.versioning import (
    AFTER_ACTION_VIEW_MODEL_SCHEMA_VERSION,
    DECISION_REVIEW_SCHEMA_VERSION,
    MISSED_EVIDENCE_ITEM_SCHEMA_VERSION,
    RUN_COMPARISON_SCORE_SCHEMA_VERSION,
    RUN_SCORE_SCHEMA_VERSION,
    SCORE_COMPONENT_SCHEMA_VERSION,
    SCORE_EXPLANATION_SCHEMA_VERSION,
    SCORE_EXPORT_ARTIFACT_SCHEMA_VERSION,
    SCORE_PROVENANCE_SCHEMA_VERSION,
    SCORE_RUBRIC_SCHEMA_VERSION,
    VALID_ALTERNATIVE_SCHEMA_VERSION,
    assert_supported_schema_version,
)

GRADING_ENGINE_VERSION = "1.0.0-phase29"


class ScoreErrorCode(StrEnum):
    SCORE_INCOMPLETE_RUN = "SCORE_INCOMPLETE_RUN"
    SCORE_RUBRIC_MISSING = "SCORE_RUBRIC_MISSING"
    SCORE_RUBRIC_INCOMPATIBLE = "SCORE_RUBRIC_INCOMPATIBLE"
    SCORE_INPUT_TAMPERED = "SCORE_INPUT_TAMPERED"
    SCORE_CHECKSUM_MISMATCH = "SCORE_CHECKSUM_MISMATCH"
    SCORE_NOT_FOUND = "SCORE_NOT_FOUND"
    SCORE_COMPARISON_INVALID = "SCORE_COMPARISON_INVALID"
    SCORE_EXPORT_FAILED = "SCORE_EXPORT_FAILED"
    SCORE_VALIDATION_FAILED = "SCORE_VALIDATION_FAILED"


class ScoreGradeBandV1(StrEnum):
    A = "A"
    B = "B"
    C = "C"
    D = "D"
    F = "F"


class DecisionReviewOutcomeV1(StrEnum):
    CREDITED = "credited"
    PENALIZED = "penalized"
    NEUTRAL = "neutral"
    RESTRAINT_CREDITED = "restraint_credited"


class ValidAlternativeKindV1(StrEnum):
    COUNTERFACTUAL = "counterfactual"
    VALID_RESPONSE_BRANCH = "valid_response_branch"
    ALTERNATIVE_DECISION = "alternative_decision"


class ScoreExportFormatV1(StrEnum):
    JSON = "json"
    MARKDOWN = "markdown"


class DecisionReviewKindV1(StrEnum):
    APPROVAL = "approval"
    REJECTION = "rejection"
    MODIFICATION = "modification"
    CANCELLATION = "cancellation"
    AGENT_RECOMMENDATION = "agent_recommendation"


class AfterActionBookmarkKindV1(StrEnum):
    DETECTION = "detection"
    EVIDENCE = "evidence"
    DECISION = "decision"
    MISSED_EVIDENCE = "missed_evidence"
    ALTERNATIVE = "alternative"
    CUSTOM = "custom"


class ScoreExplanationV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    rule_id: str = Field(alias="ruleId", min_length=1, max_length=128)
    reason: str = Field(min_length=1, max_length=2048)
    event_ids: list[EventId] = Field(default_factory=list, alias="eventIds")
    evidence_ids: list[EvidenceId] = Field(default_factory=list, alias="evidenceIds")
    decision_ids: list[str] = Field(default_factory=list, alias="decisionIds")
    hypothesis_ids: list[HypothesisId] = Field(default_factory=list, alias="hypothesisIds")
    proposal_ids: list[ProposalId] = Field(default_factory=list, alias="proposalIds")
    sequence: Sequence | None = None

    @model_validator(mode="after")
    def validate_schema_version(self) -> ScoreExplanationV1:
        assert_supported_schema_version("score_explanation", self.schema_version)
        if self.schema_version != SCORE_EXPLANATION_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported score explanation schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ScoreRubricCriterionV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    id: str = Field(min_length=1, max_length=128)
    label: str = Field(min_length=1, max_length=256)
    weight: float = Field(ge=0.0, le=1.0)
    description: str = Field(default="", max_length=1024)


class ScoreRubricV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    scenario_id: str = Field(alias="scenarioId", min_length=1, max_length=128)
    scenario_version: str = Field(alias="scenarioVersion", min_length=1, max_length=64)
    rubric_version: str = Field(alias="rubricVersion", min_length=1, max_length=64)
    grading_engine_version: str = Field(alias="gradingEngineVersion", min_length=1, max_length=64)
    max_score: float = Field(alias="maxScore", gt=0.0)
    criteria: list[ScoreRubricCriterionV1] = Field(min_length=1)
    rubric_text: str = Field(default="", alias="rubricText", max_length=8192)
    rubric_text_hash: str = Field(alias="rubricTextHash", min_length=1, max_length=128)

    @model_validator(mode="after")
    def validate_schema_version(self) -> ScoreRubricV1:
        assert_supported_schema_version("score_rubric", self.schema_version)
        if self.schema_version != SCORE_RUBRIC_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported score rubric schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ScoreComponentV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    criterion_id: str = Field(alias="criterionId", min_length=1, max_length=128)
    label: str = Field(min_length=1, max_length=256)
    weight: float = Field(ge=0.0, le=1.0)
    raw_score: float = Field(alias="rawScore", ge=0.0, le=1.0)
    weighted_contribution: float = Field(alias="weightedContribution", ge=0.0)
    rule_ids: list[str] = Field(default_factory=list, alias="ruleIds")
    explanations: list[ScoreExplanationV1] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_schema_version(self) -> ScoreComponentV1:
        assert_supported_schema_version("score_component", self.schema_version)
        if self.schema_version != SCORE_COMPONENT_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported score component schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ScoreProvenanceV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    run_id: RunId = Field(alias="runId")
    scenario_id: str = Field(alias="scenarioId", min_length=1, max_length=128)
    scenario_version: str = Field(alias="scenarioVersion", min_length=1, max_length=64)
    rubric_version: str = Field(alias="rubricVersion", min_length=1, max_length=64)
    grading_engine_version: str = Field(alias="gradingEngineVersion", min_length=1, max_length=64)
    input_event_sequence_from: Sequence = Field(alias="inputEventSequenceFrom")
    input_event_sequence_to: Sequence = Field(alias="inputEventSequenceTo")
    input_checksum: str = Field(alias="inputChecksum", min_length=1, max_length=128)
    integrity_checksum: str = Field(alias="integrityChecksum", min_length=1, max_length=128)
    calculated_at: UtcTimestamp = Field(alias="calculatedAt")
    fingerprint: str = Field(min_length=1, max_length=128)

    @model_validator(mode="after")
    def validate_schema_version(self) -> ScoreProvenanceV1:
        assert_supported_schema_version("score_provenance", self.schema_version)
        if self.schema_version != SCORE_PROVENANCE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported score provenance schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class DecisionReviewV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    decision_id: str = Field(alias="decisionId", min_length=1, max_length=128)
    kind: DecisionReviewKindV1
    sequence: Sequence
    proposal_id: ProposalId | None = Field(default=None, alias="proposalId")
    operator_action: str = Field(alias="operatorAction", min_length=1, max_length=256)
    agent_recommendation: str | None = Field(
        default=None, alias="agentRecommendation", max_length=1024
    )
    available_info_summary: str = Field(default="", alias="availableInfoSummary", max_length=2048)
    available_event_ids: list[EventId] = Field(default_factory=list, alias="availableEventIds")
    available_evidence_ids: list[EvidenceId] = Field(
        default_factory=list, alias="availableEvidenceIds"
    )
    outcome: DecisionReviewOutcomeV1
    score_delta: float = Field(alias="scoreDelta")
    rule_ids: list[str] = Field(default_factory=list, alias="ruleIds")
    explanations: list[ScoreExplanationV1] = Field(default_factory=list)
    used_future_knowledge: Literal[False] = Field(default=False, alias="usedFutureKnowledge")

    @model_validator(mode="after")
    def validate_schema_version(self) -> DecisionReviewV1:
        assert_supported_schema_version("decision_review", self.schema_version)
        if self.schema_version != DECISION_REVIEW_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported decision review schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class MissedEvidenceItemV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    expected_evidence_key: str = Field(alias="expectedEvidenceKey", min_length=1, max_length=256)
    event_type: str = Field(alias="eventType", min_length=1, max_length=128)
    asset_id: AssetId | None = Field(default=None, alias="assetId")
    description: str = Field(default="", max_length=1024)
    discovered: bool
    related_event_ids: list[EventId] = Field(default_factory=list, alias="relatedEventIds")
    related_evidence_ids: list[EvidenceId] = Field(default_factory=list, alias="relatedEvidenceIds")
    bookmark_sequence: Sequence | None = Field(default=None, alias="bookmarkSequence")

    @model_validator(mode="after")
    def validate_schema_version(self) -> MissedEvidenceItemV1:
        assert_supported_schema_version("missed_evidence_item", self.schema_version)
        if self.schema_version != MISSED_EVIDENCE_ITEM_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported missed evidence item schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ValidAlternativeV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    alternative_id: str = Field(alias="alternativeId", min_length=1, max_length=128)
    kind: ValidAlternativeKindV1
    authoritative: Literal[False] = False
    label: str = Field(min_length=1, max_length=256)
    description: str = Field(default="", max_length=2048)
    projected_overall_score: float = Field(alias="projectedOverallScore", ge=0.0)
    score_delta: float = Field(alias="scoreDelta")
    rule_ids: list[str] = Field(default_factory=list, alias="ruleIds")
    explanations: list[ScoreExplanationV1] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_schema_version(self) -> ValidAlternativeV1:
        assert_supported_schema_version("valid_alternative", self.schema_version)
        if self.schema_version != VALID_ALTERNATIVE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported valid alternative schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class RunScoreV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    score_id: str = Field(alias="scoreId", min_length=1, max_length=64)
    run_id: RunId = Field(alias="runId")
    incident_id: IncidentId | None = Field(default=None, alias="incidentId")
    overall_score: float = Field(alias="overallScore", ge=0.0)
    max_score: float = Field(alias="maxScore", gt=0.0)
    grade: ScoreGradeBandV1
    passed: bool
    components: list[ScoreComponentV1] = Field(min_length=1)
    provenance: ScoreProvenanceV1
    decision_reviews: list[DecisionReviewV1] = Field(default_factory=list, alias="decisionReviews")
    missed_evidence: list[MissedEvidenceItemV1] = Field(
        default_factory=list, alias="missedEvidence"
    )
    valid_alternatives: list[ValidAlternativeV1] = Field(
        default_factory=list, alias="validAlternatives"
    )
    coaching_text: str | None = Field(default=None, alias="coachingText", max_length=8192)
    coaching_authoritative: Literal[False] = Field(default=False, alias="coachingAuthoritative")
    hidden_cause_id: str | None = Field(default=None, alias="hiddenCauseId", max_length=128)
    hidden_cause_label: str | None = Field(default=None, alias="hiddenCauseLabel", max_length=256)
    hidden_cause_revealed: bool = Field(default=False, alias="hiddenCauseRevealed")

    @model_validator(mode="after")
    def validate_schema_version(self) -> RunScoreV1:
        assert_supported_schema_version("run_score", self.schema_version)
        if self.schema_version != RUN_SCORE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported run score schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class AfterActionBookmarkRefV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    label: str = Field(min_length=1, max_length=256)
    sequence: Sequence
    kind: AfterActionBookmarkKindV1
    related_id: str | None = Field(default=None, alias="relatedId", max_length=128)


class AfterActionTimelineHighlightV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    sequence: Sequence
    label: str = Field(min_length=1, max_length=512)
    kind: str = Field(min_length=1, max_length=64)
    related_id: str | None = Field(default=None, alias="relatedId", max_length=128)


class AfterActionViewModelV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    run_id: RunId = Field(alias="runId")
    run_status: str = Field(alias="runStatus", min_length=1, max_length=64)
    score: RunScoreV1
    affected_asset_ids: list[AssetId] = Field(default_factory=list, alias="affectedAssetIds")
    lessons: list[str] = Field(default_factory=list)
    scribe_report_id: str | None = Field(default=None, alias="scribeReportId", max_length=64)
    scribe_version_number: int | None = Field(default=None, alias="scribeVersionNumber", ge=1)
    bookmarks: list[AfterActionBookmarkRefV1] = Field(default_factory=list)
    timeline_highlights: list[AfterActionTimelineHighlightV1] = Field(
        default_factory=list, alias="timelineHighlights"
    )

    @model_validator(mode="after")
    def validate_schema_version(self) -> AfterActionViewModelV1:
        assert_supported_schema_version("after_action_view_model", self.schema_version)
        if self.schema_version != AFTER_ACTION_VIEW_MODEL_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported after-action view model schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class RunComparisonComponentDeltaV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    criterion_id: str = Field(alias="criterionId", min_length=1, max_length=128)
    left_raw_score: float = Field(alias="leftRawScore", ge=0.0, le=1.0)
    right_raw_score: float = Field(alias="rightRawScore", ge=0.0, le=1.0)
    delta: float


class RunComparisonV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    left_run_id: RunId = Field(alias="leftRunId")
    right_run_id: RunId = Field(alias="rightRunId")
    left_score_id: str = Field(alias="leftScoreId", min_length=1, max_length=64)
    right_score_id: str = Field(alias="rightScoreId", min_length=1, max_length=64)
    left_overall_score: float = Field(alias="leftOverallScore", ge=0.0)
    right_overall_score: float = Field(alias="rightOverallScore", ge=0.0)
    overall_delta: float = Field(alias="overallDelta")
    component_deltas: list[RunComparisonComponentDeltaV1] = Field(
        default_factory=list, alias="componentDeltas"
    )
    grading_engine_version: str = Field(alias="gradingEngineVersion", min_length=1, max_length=64)

    @model_validator(mode="after")
    def validate_schema_version(self) -> RunComparisonV1:
        assert_supported_schema_version("run_comparison", self.schema_version)
        if self.schema_version != RUN_COMPARISON_SCORE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported run comparison schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ScoreExportArtifactV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    export_id: str = Field(alias="exportId", min_length=1, max_length=64)
    score_id: str = Field(alias="scoreId", min_length=1, max_length=64)
    run_id: RunId = Field(alias="runId")
    format: ScoreExportFormatV1
    scenario_version: str = Field(alias="scenarioVersion", min_length=1, max_length=64)
    rubric_version: str = Field(alias="rubricVersion", min_length=1, max_length=64)
    grading_engine_version: str = Field(alias="gradingEngineVersion", min_length=1, max_length=64)
    integrity_checksum: str = Field(alias="integrityChecksum", min_length=1, max_length=128)
    content_checksum: str = Field(alias="contentChecksum", min_length=1, max_length=128)
    created_at: UtcTimestamp = Field(alias="createdAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> ScoreExportArtifactV1:
        assert_supported_schema_version("score_export_artifact", self.schema_version)
        if self.schema_version != SCORE_EXPORT_ARTIFACT_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported score export artifact schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self
